### PR TITLE
[FIX](struct) fix struct be nested when read will make core

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx
 1. fix struct be nested when read will make BE core 
2. fix old planner do not support struct be nested 

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 1]
```diff
@@ -787,9 +788,49 @@ public static Type getAssignmentCompatibleType(Type t1, Type t2, boolean strict)
...
 
             return new ArrayType(itemCompatibleType, arrayType1.getContainsNull() || arrayType2.getContainsNull());
-        } else if (t1.isArrayType() && t2.isNull()) {
+        } else if (t1.isMapType() && t2.isMapType()) {
+            MapType mapType1 = (MapType) t1;
+            MapType mapType2 = (MapType) t2;
+            Type keyCompatibleType = Type.getAssignmentCompatibleType(mapType1.getKeyType(),
+                    mapType2.getKeyType(), strict);
+            if (keyCompatibleType.isInvalid()) {
+                return keyCompatibleType;
```
Comment (morrySnow) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:797]: why return keyCompatibleType here?

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 2]
```diff
@@ -2169,6 +2210,17 @@ public static boolean matchExactType(Type type1, Type type2, boolean ignorePreci
                     return false;
                 }
                 return matchExactType(((ArrayType) type2).getItemType(), ((ArrayType) type1).getItemType());
+            } else if (type2.isMapType()) {
+                // For types array, we also need to check contains null for case like
+                // cast(array<not_null(int)> as array<int>)
+                if (!((MapType) type2).getIsKeyContainsNull() == ((MapType) type1).getIsKeyContainsNull()) {
```
Comment (morrySnow) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:2216]:
```suggestion
                if (((MapType) type2).getIsKeyContainsNull() != ((MapType) type1).getIsKeyContainsNull()) {
```

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 3]
```diff
@@ -2169,6 +2210,17 @@ public static boolean matchExactType(Type type1, Type type2, boolean ignorePreci
...
+                    return false;
+                }
+                if (!((MapType) type2).getIsValueContainsNull() == ((MapType) type1).getIsValueContainsNull()) {
```
Comment (morrySnow) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:2219]:
```suggestion
                if (((MapType) type2).getIsValueContainsNull() != ((MapType) type1).getIsValueContainsNull()) {
```

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java (hunk 1 above)
Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:797]: here keyCompatibleType equals Invalid Type, so just return InvalidType

**Comment by amorynan** [comment]
run buildall

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java (hunk 3 above)
Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java]: done!

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java (hunk 2 above)
Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java]: done!

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java (hunk 1 above)
Comment (amorynan) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java]: done!

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run feut

**Comment by amorynan** [comment]
run beut

**Comment by amorynan** [comment]
run feut

**Comment by amorynan** [comment]
run p0

**Review by morrySnow — DISMISSED** [review]

**Review by xiaokang — DISMISSED** [review]
LGTM

**Comment by amorynan** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**